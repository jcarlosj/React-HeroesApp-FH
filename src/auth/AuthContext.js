import { createContext } from 'react';

/** AuthContext: High Order Component */
export const AuthContext = createContext();