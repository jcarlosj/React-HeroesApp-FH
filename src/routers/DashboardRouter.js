import React from "react";
import {
  Switch,
  Route,
  Redirect
} from "react-router-dom";

/** Components */
import { NavBar } from '../components/ui/NavBar';
import { MarvelPage } from "../components/marvel/MarvelPage";
import { DcPage } from "../components/dc/DcPage";
import { HeroPage } from "../components/heroes/HeroPage";

/** Secondary Router: Functional Component */
export const DashboardRouter = () => {
    return (
        <>
            <NavBar />
            <div>

                {/* A <Switch> looks through its children <Route>s and
                    renders the first one that matches the current URL. */}
                <Switch>

                    <Route exact path="/dc" component={ DcPage } />
                    <Route exact path="/marvel" component={ MarvelPage } />
                    <Route exact path="/hero/:hero_id" component={ HeroPage } />
                    <Redirect to path="/marvel" />
                        
                </Switch>
            </div>
        </>
    )
}
