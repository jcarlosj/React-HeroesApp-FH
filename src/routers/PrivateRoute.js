import React from 'react';
import { Redirect, Route } from 'react-router-dom';
import PropTypes from 'prop-types';

/** Funcional Component */
export const PrivateRoute = ({
    isAuthenticated,                        //  Propiedad definida esperada
    component: Component,                   //  Propiedad definida esperada
    ...rest                                 //  Propiedades indefinidas esperadas
}) => {
    return (
        <Route
            { ...rest }                             //  Pasa path para definir la ruta
            component={ props => (          
                ( isAuthenticated )                 //  Condiciona el componente a pasar a la ruta 
                    ?   <Component { ...props } />  //  Pasa los props requeridos por el componente
                    :   <Redirect to="/login" />    //  Redirecciona en caso de no tener acceso permitido
            )}
        />
    )
}

/** Implementamos PropTypes para manejar errores de implementaci√≥n */
PrivateRoute.propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired
}