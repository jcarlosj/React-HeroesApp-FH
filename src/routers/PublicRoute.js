import React from 'react';
import { Redirect, Route } from 'react-router-dom';
import PropTypes from 'prop-types';

/** Funcional Component */
export const PublicRoute = ({
    isAuthenticated,                        //  Propiedad definida esperada
    component: Component,                   //  Propiedad definida esperada
    ...rest                                 //  Propiedades indefinidas esperadas
}) => {
    return (
        <Route
            { ...rest }                             //  Pasa path para definir la ruta
            component={ props => (          
                ( ! isAuthenticated )                 //  Condiciona el componente a pasar a la ruta 
                    ?   <Component { ...props } />  //  Pasa los props requeridos por el componente
                    :   <Redirect to="/" />    //  Redirecciona en caso de no tener acceso permitido
            )}
        />
    )
}

/** Implementamos PropTypes para manejar errores de implementaci√≥n */
PublicRoute.propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired
}