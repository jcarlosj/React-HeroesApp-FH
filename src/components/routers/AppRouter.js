import React from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route
} from "react-router-dom";

/** Router */
import { DashboardRouter } from "./DashboardRouter";

/** Components */
import { LoginPage } from "../login/LoginPage";

/** Main Router: Functional Component */
export const AppRouter = () => {
    return (
        <Router>
            <div>

                {/* A <Switch> looks through its children <Route>s and
                    renders the first one that matches the current URL. */}
                <Switch>

                    <Route exact path="/login" component={ LoginPage } />
                    <Route path="/" component={ DashboardRouter } />
                        
                </Switch>
            </div>
        </Router>
    )
}
