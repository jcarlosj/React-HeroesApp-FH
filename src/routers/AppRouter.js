import React, { useContext } from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route
} from "react-router-dom";

/** Router */
import { DashboardRouter } from "./DashboardRouter";

/** Components */
import { LoginPage } from "../components/login/LoginPage";
import { PrivateRoute } from "./PrivateRoute";

/** Contexts */
import { AuthContext } from "../auth/AuthContext";

/** Main Router: Functional Component */
export const AppRouter = () => {

    const { user: { logged } } = useContext( AuthContext );

    return (
        <Router>
            <div>

                {/* A <Switch> looks through its children <Route>s and
                    renders the first one that matches the current URL. */}
                <Switch>

                    <Route exact path="/login" component={ LoginPage } />
                    <PrivateRoute 
                        isAuthenticated={ logged }      //  Autenticacion (true/false): Propiedad definida esperada
                        component={ DashboardRouter }   //  Router: Propiedad definida esperada
                        path="/"                        //  Ruta para la que se define el acceso privado (Esto afectara a todas las rutas hijas del Router)
                    />
                        
                </Switch>
            </div>
        </Router>
    )
}
