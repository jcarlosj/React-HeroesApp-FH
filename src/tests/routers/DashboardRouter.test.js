import '@testing-library/jest-dom';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

/** Context */
import { AuthContext } from '../../auth/AuthContext';

/** Router */
import { DashboardRouter } from '../../routers/DashboardRouter';

describe( 'Dashboard Router', () => {

    const contextValue = {      
        user: {                 //  State del Reducer
            id: expect.any( Number ),
            name: "Luisa Maria", 
            email: "luisa.maria@correo.co",
            logged: true
        },                           
        dispatch: jest.fn()     //  Disparador de acciones de nuestro Context
    }

    test( 'debe mostrarse correctamente', () => { 
    
        /** Esta prueba evalua al componente hijo, para ello se debe usar 'mount' para montar el componente, en el momento esta funcion no es compatible con React 17 */
        const wrapper = mount( 
            <AuthContext.Provider value={       //  Hace disponible las propiedades y funciones del reducer en todo el árbol de componentes de la aplicación
                contextValue    
            }>
                <MemoryRouter>
                    <DashboardRouter /> 
                </MemoryRouter>
            </AuthContext.Provider>
        );
        
        // console.log( wrapper.debug() );
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find( 'span.text-white' ).text().trim() ).toBe( 'Hola, Luisa Maria' );

    } );

} );