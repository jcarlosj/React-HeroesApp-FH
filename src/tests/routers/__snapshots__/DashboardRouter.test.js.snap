// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Router debe mostrarse correctamente 1`] = `
<Fragment>
  <NavBar />
  <div>
    <Switch>
      <Route
        component={[Function]}
        exact={true}
        path="/dc"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/marvel"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/search"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/hero/:hero_id"
      />
      <Redirect
        to="/marvel"
      />
    </Switch>
  </div>
</Fragment>
`;
